common/transport/amqp/src/receiver_link.ts

```diff
@@ -80,38 +115,61 @@ export class ReceiverLink  extends EventEmitter implements AmqpLink {
+              debug('In the error handler for the attaching receiver');
+              manageAttachingReceiverHandlers('removeListener');
+              this._fsm.transition('detached', callback, context);
+            };
+            const attachingReceiverDisconnectedHandler = (context) => {
+              debug('In the disconnected handler for the attaching receiver');
+              manageAttachingReceiverHandlers('removeListener');
+              this._fsm.transition('detached', callback, context);
+            };
+            const attachingReceiverCloseHandler = (context) => {
+              debug('In the close handler for attaching receiver');
+              manageAttachingReceiverHandlers('removeListener');
+              this._fsm.transition('detached', callback, context);
+            };
+            const attachingReceiverCancelHandler = (context) => {
+              debug('In the cancel handler for attaching receiver');
+              manageAttachingReceiverHandlers('removeListener');
+              this._fsm.transition('detached', callback, context);
+            };
+            const attachingReceiverOpenHandler = (context) => {
+              debug('In the open handler for attaching receiver');
+              this._linkObject = context.receiver;
+              manageAttachingReceiverHandlers('removeListener');
+              manageAttachedReceiverHandlers('on');
+              this._fsm.transition('attached', callback);
+            };
+            manageAttachingReceiverHandlers = (operation: string) => {
+              if (this._amqpSession) {
+                this._amqpSession[operation]('cancelAzureIotSdkAmqpAttaching', attachingReceiverCancelHandler);
+                this._amqpSession[operation]('receiver_open', attachingReceiverOpenHandler);
+                this._amqpSession[operation]('receiver_close', attachingReceiverCloseHandler);
+                this._amqpSession[operation]('disconnected', attachingReceiverDisconnectedHandler);
+                this._amqpSession[operation]('error', attachingReceiverErrorHandler);
+              } else {
+                debug('manage receiver attaching state handlers called with no session object');
+              }
+            };
+            manageAttachingReceiverHandlers('on');
+            /*Codes_SRS_NODE_COMMON_AMQP_16_018: [The `attachReceiverLink` method shall call `createReceiver` on the `amqp10` client object.]*/
+            this._amqpSession.open_receiver(this._combinedOptions);
+          },
+          attach: (null),
+          detach: (callback) => {
+            this._amqpSession.emit('cancelAzureIotSdkAmqpAttaching');
+            this._fsm.transition('detached', callback, new Error('detached while attaching'));
           },
-          detach: (callback) => this._fsm.transition('detaching', callback),
           forceDetach: (err) => {
+            this._amqpSession.emit('cancelAzureIotSdkAmqpAttaching');
```

Comment:
>_amqpSession [](start = 17, length = 12)

that seems sketchy. in principle, we're not supposed to modify the behavior of an object of a dependency